pkg/planner/core/tiflash_predicate_push_down.go

```diff
@@ -0,0 +1,370 @@
+	tableRowCount := ts.StatsInfo().RowCount
+
+	for _, exprGroup := range sortedConds {
+		mergedConds := append(selectedConds, exprGroup.exprs...)
+		selectivity, _, err := cardinality.Selectivity(sctx, ts.tblColHists, mergedConds, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		colCnt := expression.ExtractColumnSet(mergedConds...).Len()
+		income := (1 - selectivity) * tableRowCount
+		// If selectedColumnCount does not change,
+		// or the increase of the number of filtered rows is greater than tiflashDataPackSize and the income increases, push down the conditions.
+		if colCnt == selectedColumnCount || (income > tiflashDataPackSize && income*(float64(totalColumnCount)-float64(colCnt)) > selectedIncome) {
+			selectedConds = mergedConds
+			selectedColumnCount = colCnt
+			selectedIncome = income * (float64(totalColumnCount) - float64(colCnt))
+			selectedSelectivity = selectivity
+		} else if income < tiflashDataPackSize {
+			// If the increase of the number of filtered rows is less than tiflashDataPackSize,
+			// break the loop to reduce the cost of calculating selectivity.
+			break
+		}
+	}
+
+	if len(selectedConds) == 0 {
+		return
+	}
+	// add the pushed down conditions to table scan
+	ts.LateMaterializationFilterCondition = selectedConds
+	ts.lateMaterializationSelectivity = selectedSelectivity
+}
+
+// isPredicateSimpleCompare is used to check if the condition is a simple comparison predicate
+// which only contains >, >=, =, !=, <=, <, in.
+func isPredicateSimpleCompare(cond expression.Expression) bool {
+	if sf, ok := cond.(*expression.ScalarFunction); ok {
+		switch sf.FuncName.L {
+		case ast.EQ, ast.GE, ast.LE, ast.LT, ast.GT, ast.In:
+			return true
+		case ast.LogicAnd, ast.LogicOr:
+			return isPredicateSimpleCompare(sf.GetArgs()[0]) && isPredicateSimpleCompare(sf.GetArgs()[1])
+		case ast.UnaryNot:
+			return isPredicateSimpleCompare(sf.GetArgs()[0])
+		}
+	}
+	return false
+}
+
+// handleTiFlashPredicatePushDown is used to handle the TiFlash predicate push down.
+// 1. Whether to use the inverted index.
+// 2. Whether to push down the conditions to the table scan.
+func handleTiFlashPredicatePushDown(pctx base.PlanContext, ts *PhysicalTableScan, indexHints []*ast.IndexHint) {
+	// Consider use index hints
+	indexMap := make(map[string]int, len(ts.Table.Indices))
+	for _, hint := range indexHints {
+		if hint.HintScope != ast.HintForScan {
+			continue
+		}
+
+		switch hint.HintType {
+		case ast.HintUse:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]++
+				}
+			}
+		case ast.HintIgnore:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]--
+				}
+			}
+		case ast.HintForce:
+			for _, name := range hint.IndexNames {
+				indexMap[name.L] = math.MaxInt
+			}
+		default:
+			continue
+		}
+	}
```

Comment:
We can store the indexes filtered by the hint in datasource.
Directly add a field like `UsedColumnarIndexes` for `DataSource` instead of a field `AstIndexHints`.

Reply:
Left it in another PR.